## ALGUNOS EJEMPLOS DE ALGEBRA MATRICIAL EN JULIA
### MECÁNICA DE SÓLIDOS
### Ramón Zúñiga, CGEO UNAM

In [2]:
using Pkg

In [3]:
Pkg.add("LinearAlgebra")

    Updating registry at `~/.julia/registries/General`
┌ Info: The General registry is installed via git. Consider reinstalling it via
│ the newer faster direct from tarball format by running:
│   pkg> registry rm General; registry add General
└ 
    Updating git-repo `https://github.com/JuliaRegistries/General.git`
   Resolving package versions...
  No Changes to `~/.julia/environments/v1.11/Project.toml`
  No Changes to `~/.julia/environments/v1.11/Manifest.toml`


In [4]:
using LinearAlgebra

Supongamos una matriz

\begin{equation}\boldsymbol A =
\begin{bmatrix} 4 & -2  \\
                1 & 1  \\
\end{bmatrix}\end{equation}



In [5]:
A = [
    4 -2
    1 1
]

2×2 Matrix{Int64}:
 4  -2
 1   1

In [6]:
inv(A) # Inversa de la Matriz A

2×2 Matrix{Float64}:
  0.166667  0.333333
 -0.166667  0.666667

Confirmamos que

$ A \cdot A^{-1} = I $

In [115]:
A*inv(A)

2×2 Matrix{Float64}:
 1.0          0.0
 2.77556e-17  1.0

In [116]:
A' # Adjunta de la matriz A

2×2 adjoint(::Matrix{Int64}) with eltype Int64:
  4  1
 -2  1

In [7]:
det(A) # Determinante de A

6.0

In [118]:
transpose(A)  # Transpuesta de A

2×2 transpose(::Matrix{Int64}) with eltype Int64:
  4  1
 -2  1

### Valores Propios y Vectores Propios (Eigenvalores y Eigenvectores)

Los valores propios λ y los vectores propios v
  son las soluciones de
    
\begin{equation}
(\boldsymbol{A}-\mathbf{I}\lambda)\mathbf{v}
\tag{5} = 0
\end{equation}

siendo A la matriz a la que se requiere obtener sus valores propios

In [119]:
λ = eigvals(A)

2-element Vector{Float64}:
 2.0
 3.0

In [120]:
v = eigvecs(A)

2×2 Matrix{Float64}:
 0.707107  0.894427
 0.707107  0.447214

Lo que sigue es para confirmar la operación fundamental de la matrix con un eigenvector que debe dar:

\begin{equation}
\boldsymbol{\sigma}\mathbf{v}=\lambda\mathbf{v}
\end{equation}

In [121]:
A*v[:,1]  # Esto multiplica la matriz A por el primer eigenvector

2-element Vector{Float64}:
 1.414213562373095
 1.414213562373095

Notar que los dos valores anteriores son únicamente los valores del eigenvector por el primer eigenvalor
0.707107*2.0  # los dos valores del eigenvector 1 son iguales

In [123]:
A*v[:,2]  # Ahora multiplicamos por el segundo eigenvector

2-element Vector{Float64}:
 2.6832815729997477
 1.3416407864998738

In [124]:
# en el caso del segundo eigenvector tenemos dos valores diferentes, pero el resultado es igual a multiplicarlos
# por el segundo eigenvalor

In [125]:
0.894427*3

2.683281

In [126]:
0.447214*3

1.341642

Vamos a generar una matriz aleatoria de 2 x 1

In [8]:
B = rand(2,1)* 5

2×1 Matrix{Float64}:
 1.7337131710313542
 1.3841728910830704

Multiplicación

In [128]:
C = A*B

2×1 Matrix{Float64}:
 2.577245618451669
 1.3117659523428489

In [129]:
D = B*A # aquí confirmamos que la multiplicación no es conmutativa

LoadError: DimensionMismatch: matrix A has dimensions (2,1), matrix B has dimensions (2,2)

In [130]:
E = rand(2,2)

2×2 Matrix{Float64}:
 0.713446   0.518419
 0.0931893  0.80004

In [131]:
F = A*E

2×2 Matrix{Float64}:
 2.66741   0.473594
 0.806635  1.31846

In [132]:
G = E*A

2×2 Matrix{Float64}:
 3.3722  -0.908473
 1.1728   0.613661

#### Escritura y Lectura de Datos Tabulares
#### Creamos datos (matriz A), los guardamos en un archivo con el nombre de DatosA especificando espaciador con blancos, y después lo volvemos a leer como la matriz B

In [9]:
Pkg.add("DelimitedFiles")

   Resolving package versions...
    Updating `~/.julia/environments/v1.11/Project.toml`
  [8bb1440f] + DelimitedFiles v1.9.1
  No Changes to `~/.julia/environments/v1.11/Manifest.toml`


In [133]:
using DelimitedFiles
A = rand(50,2) # Creamos datos al azar, por ejemplo, una matriz de 50 filas y dos columnas

writedlm("DatosA", A, ' ')  # Escribimos los datos con espacios en un archivo de nombre DatosA 
B = readdlm("DatosA", ' ');   # Leemos los datos del archivo indicando que los espacios son separadores


---
## Ejercicios Propuestos

**Ej. 1 — Otra matriz 3×3.**  
Sea $\boldsymbol{A} = \begin{bmatrix} 6 & 2 & 1 \\ 2 & 5 & 3 \\ 1 & 3 & 4 \end{bmatrix}$. Calcula su inversa, determinante y transpuesta. Verifica que $\boldsymbol{A}\boldsymbol{A}^{-1} = \boldsymbol{I}$.

---
**Ej. 2 — Esfuerzos y direcciones principales.**  
Para el tensor $\boldsymbol{\sigma} = \begin{bmatrix} 15 & 3 & 0 \\ 3 & 8 & 2 \\ 0 & 2 & 12 \end{bmatrix}$ MPa, calcula los eigenvalores (esfuerzos principales) y eigenvectores (direcciones principales). Verifica la relación $\boldsymbol{\sigma}\,\mathbf{v}_i = \lambda_i\,\mathbf{v}_i$ para cada par.

---
**Ej. 3 — No conmutatividad.**  
Usando $\boldsymbol{A} = \begin{bmatrix} 1 & 2 & 3 \\ 0 & 1 & 4 \\ 5 & 6 & 0 \end{bmatrix}$ y $\boldsymbol{B} = \begin{bmatrix} 3 & 1 & 2 \\ 4 & 0 & 1 \\ 2 & 3 & 1 \end{bmatrix}$, calcula $\boldsymbol{A}\boldsymbol{B}$ y $\boldsymbol{B}\boldsymbol{A}$. ¿Son iguales?

---
**Ej. 4 — Determinante y eigenvalores.**  
Verifica que $\det(\boldsymbol{\sigma}) = \lambda_1 \cdot \lambda_2 \cdot \lambda_3$ para el tensor del Ej. 2.  
> ¿Qué significaría físicamente $\det(\boldsymbol{\sigma}) = 0$?

In [ ]:
using LinearAlgebra

# ── Ejercicio 1 ────────────────────────────────────────────────────────────────
A = [6 2 1; 2 5 3; 1 3 4]
println("=== Ejercicio 1 ===")
println("Inversa:");           display(inv(A))
println("Verif. A·A⁻¹ = I:"); display(round.(A * inv(A), digits=10))
println("det(A) = ",          det(A))
println("Transpuesta:");       display(A')

# ── Ejercicio 2 ────────────────────────────────────────────────────────────────
σ = [15 3 0; 3 8 2; 0 2 12]
println("\n=== Ejercicio 2 ===")
λ = eigvals(σ)
v = eigvecs(σ)
println("Esfuerzos principales (MPa): ", round.(λ, digits=4))
println("Direcciones principales (columnas):")
display(round.(v, digits=4))
println("\nVerificación σ·vᵢ = λᵢ·vᵢ  (error debe ser ≈ 0):")
for i in 1:3
    println("  Error eigenvector $i: ", norm(σ * v[:,i] - λ[i] * v[:,i]))
end
println("σ_max = ", round(maximum(λ), digits=4), " MPa → dirección: ", round.(v[:, argmax(λ)], digits=4))

# ── Ejercicio 3 ────────────────────────────────────────────────────────────────
A = [1 2 3; 0 1 4; 5 6 0]
B = [3 1 2; 4 0 1; 2 3 1]
println("\n=== Ejercicio 3 ===")
println("A·B - B·A:"); display(A*B - B*A)
println("¿A·B == B·A? ", (A*B) ≈ (B*A))
# TODO: ¿bajo qué condición especial sí conmutan dos matrices?

# ── Ejercicio 4 ────────────────────────────────────────────────────────────────
σ = [15 3 0; 3 8 2; 0 2 12]
λ = eigvals(σ)
println("\n=== Ejercicio 4 ===")
println("det(σ)   = ", det(σ))
println("∏ λᵢ     = ", prod(λ))
println("¿Iguales? ", det(σ) ≈ prod(λ))
# TODO: modifica σ para que un eigenvalor sea cero. ¿Qué le pasa a det(σ)?